lib/internal/fs/copy/copy.js

```diff
@@ -0,0 +1,468 @@
+'use strict';
```

Comment:
Could this file be refactored to use async/await instead of callbacks? IMHO it would improve greatly the readability.

Reply:
@aduh95 the thought had crossed my mind too. My reasoning around keeping it `cb` based was as follows:

1. It keeps us closer to the source material, if new bugs are found in `fs-extra`, or in `node`, we can share the fix between the two codebases more easily.
2. Porting the algorithm would be a pretty significant refactor, and I'm worried we might lose some of the benefit of using a battle-tested implementation.

Perhaps this could be a future refactor after we have really solid code coverage\, and understand all the existing behavior?

Reply:
The upside of doing this refactor now rather than later is we won't break anyone's workflow; if we first release a callback-based version, and when switching to promises the behavior changes, it will break someone workflow.
I'm -0 landing a callback-centered API, and I volunteer to do the refactor if that helps you (I can open a PR to your branch if you want).
Anyway, I agree this should wait for better test coverage though, no matter what we decide on this subject.

Reply:
@aduh95 I'd be open to the refactor from your branch, once I get test coverage improved 👍 we're at about 77% right now.

I would be interested to here @manidlou and @RyanZim's thoughts on this too, are there any foot guns we should be mindful of?

Reply:
By all means, go for the refactor. We never did it, mixture of laziness and fear of breakage, but now seems to be a good time to do it.

Reply:
@aduh95 I have coverage up to 93% and I believe have test cases for all the important logical branches, if you want to take a stab at submitting a patch against this branch switching `copy.js` to using promises.

I still intend to flesh out tests for `copy-sync.js` and the promises version of API (which I will do after this refactor).

Let me know if you don't have time, and I could take a stab, otherwise I'll be waiting on your PR.

Reply:
https://github.com/bcoe/node-1/pull/1 🤞

Reply:
@aduh95 tests continued passing 🎉 Thanks for the patch.

I'll next work on getting coverage as high for sync, a few more smoke tests for the promises API, and finally getting a Windows vm running for testing.